ext/rotoscope/rotoscope.c

```diff
@@ -176,19 +239,39 @@ static void event_hook(VALUE tpval, void *data) {
   }
 
   rb_trace_arg_t *trace_arg = rb_tracearg_from_tracepoint(tpval);
+  rb_event_flag_t event_flag = rb_tracearg_event_flag(trace_arg);
   rs_callsite_t trace_path = tracearg_path(trace_arg);
 
-  if (rejected_path(trace_path.filepath, config)) return;
+  if (rejected_path(trace_path.filepath, config)) {
+    if (METHOD_CALL_P(event_flag)) return;
+    // We must ensure the event being blacklisted has correct filepath set,
+    // since block-method returns always point one level too high in the stack.
+
+    // Does this match the last caller?
+    rs_stack_frame_t *last_frame = rs_stack_peek(&config->stack);
+    if (!tracepoint_entities_match(tpval, last_frame->tp.raw)) return;
+    // Are we dealing with a block return in wrong context?
+    if (!endof_block_method(trace_arg)) return;
+  }
 
   rs_tracepoint_t trace = extract_full_tracevals(trace_arg, &trace_path);
+  trace.raw = tpval;
```

Comment:
The same TracePoint object is passed to `event_hook` as the one that called TracePoint#enable, so storing the object to compare frames will always result in the frames matching.

Reply:
🤦‍♂️ I'll have to build my own struct to store the data then, since the result of `rb_tracearg_from_tracepoint` is only a pointer that doesn't last very long AFAICT. Good catch.